src/argprocessing.cpp

```diff
@@ -1302,17 +1303,22 @@ process_args(Context& ctx)
   // Since output is redirected, compilers will not color their output by
   // default, so force it explicitly.
   nonstd::optional<std::string> diagnostics_color_arg;
-  if (config.compiler_type() == CompilerType::clang) {
+  switch (config.compiler_type()) {
+  case CompilerType::clang:
+  case CompilerType::msvc:
```

Comment:
This looks like a case where we can't let `CompilerType::msvc` do "clang-cl things" since the code as currently written will add `-fcolor-diagnostics` to the command line, which won't work for MSVC.

I suggest not support color diagnostics for clang-cl for now.

Reply:
Good point, but it's too bad not to supported.

Removed for now, but I think it will be better to have some sort of indication for clang after all.

Reply:
Perhaps I should try to explain why I'm spontaneously pushing back on this (having clang-cl as a separate compiler type). The main reason is that it's not clear to me how to represent this somewhat strange mix of two compiler types (i.e., clang-cl) when implementing #956, and I don't want to add logic now that makes such code refactoring even harder. I'd rather wait with such features.

Also, what I agreed to in https://github.com/ccache/ccache/pull/506#issuecomment-962452481 was to add basic support for MSVC, with the original patch being small and well-contained. But now there have are been several MSVC-related fixes and features on top of that and also clang-cl support, which just takes more steps on the road of accidental complexity which I have to maintain. This is demotivating to me.

Reply:
I see your point.

I don't use clang-cl personally, I just wanted to have good support for msvc.

Regarding the abstraction, I think the original patch was more accurate.

When refactoring, you can have feature sets, and each compiler can be associated with one or more of them. So gcc will have gcc-like, clang will have gcc-like + clang, and clang-cl will have msvc + clang.

Then arguments can be supported by querying the feature sets of the compiler.

This is a raw idea, and it needs more processing. What do you think? Do you have a design already?

Reply:
No, I haven't begun thinking about that since I don't have time, which is why I'm by default a bit sceptical of change right now. :raised_eyebrow: That said, I don't like the current PR version's `ctx.config.compiler_type() == CompilerType::msvc` check in `process_profiling_option` either since it's "lying".

I wanted an answer to _"What would happen if clang-cl was just characterized as cl instead?"_ and now I've got it. Let's go with the original approach for now. (No guarantees that it will stay that way in the future, though.)